drone/src/agent/engines/docker/mod.rs

```diff
@@ -316,4 +307,56 @@ impl Engine for DockerInterface {
 
         Box::pin(stream)
     }
-}
\ No newline at end of file
+
+    async fn load(&self, spawn_request: &SpawnRequest) -> Result<()> {
+        self.pull_image(
+            &spawn_request.executable.image,
+            &spawn_request
+                .executable
+                .credentials
+                .as_ref()
+                .map(|d| d.into()),
+        )
+        .await?;
+
+        let backend_id = spawn_request.backend_id.to_resource_name();
+        self.run_container(
+            &backend_id,
+            &spawn_request.executable.image,
+            &spawn_request.executable.env,
+            &spawn_request.executable.resource_limits,
+        )
+        .await?;
+        tracing::info!(%backend_id, "Container is running.");
+
+        Ok(())
+    }
+
+    async fn backend_status(&self, backend: &BackendId) -> Result<EngineBackendStatus> {
+        let container_name = backend.to_resource_name();
+        let container = match self.docker.inspect_container(&container_name, None).await {
+            Ok(container) => container,
+            Err(bollard::errors::Error::DockerResponseServerError {
+                status_code: 404, ..
+            }) => return Ok(EngineBackendStatus::Unknown),
+            Err(err) => return Err(err.into()),
+        };
+        let state = container
+            .state
+            .ok_or_else(|| anyhow!("No state found for container."))?;
+
+        let running = state
+            .running
+            .ok_or_else(|| anyhow!("State found but no running field for container."))?;
+
+        if running {
+            Ok(EngineBackendStatus::Running)
+        } else {
+            match state.exit_code {
+                None => Ok(EngineBackendStatus::Terminated),
```

Comment:
Is this correct? Don't terminated processes still have exit codes?

Reply:
I haven't tested this, but my understanding is:

- If a process receives `SIGTERM` (graceful termination) and then exits, it returns an exit code.
- If a process is killed by `SIGKILL`, it does not have an exit code.
- Docker attempts to kill a process first gracefully by sending a `SIGTERM`, then forcefully by sending a `SIGKILL` after ~10s.

The underlying problem here is that from Docker state alone, we can't infer _why_ a container was killed anyway. Fortunately, we only use this method when we are either in the `Starting` and `Ready` states; if a container is swept or terminated by Plane, it has already advanced to another state and we don't need to inquire about its status from Docker.

Reply:
It looks like the first answer here might be correct: https://stackoverflow.com/questions/31297616/what-is-the-authoritative-list-of-docker-run-exit-codes

When I tested this, `docker inspect` returned an exit code of `0` while it was still running (I guess it's a non-nullable value?) and when I ran `docker kill`, the `docker inspect` command showed the exit code as `137`.

Reply:
Good find. I think you're right that it's not a nullable value; bollard is generated from the OpenAPI spec and I don't know OpenAPI that well, but it doesn't seem to indicate that it's a required field. It seems to be a case of the field always being populated but not required by the API spec.

Unfortunately it means that a container that was killed is indistinguishable from one that exited successfully, unless that's represented elsewhere.

So `None` here is probably an impossible scenario, but we need to handle it somehow to satisfy the typechecker, and I don't think it makes sense to panic since:
- Getting a `None` here isn't particularly alarming
- Treating it as terminated seems like a reasonable thing to do

Reply:
Might make sense to add a comment or something, as it implies to the reader that this is how we can see if an container has terminated. Not a blocker though.